kratos/includes/registry_item.h

```diff
@@ -97,18 +223,30 @@ class RegistryItem
         std::string const& ItemName,
         TArgumentsList&&... Arguments)
     {
-        KRATOS_ERROR_IF(this->HasItem(ItemName)) << "The RegistryItem " << this->Name() << " already has an item with name " << ItemName << std::endl;
+        KRATOS_ERROR_IF(this->HasItem(ItemName)) << "The RegistryItem '" << this->Name() << "' already has an item with name " << ItemName << "." << std::endl;
+        KRATOS_ERROR_IF(this->HasValue()) <<
+            "Trying to add '"<< ItemName << "' item to the RegistryItem '" << this->Name() << "' but this already has value. Items cannot have both value and subitem." << std::endl;
         auto insert_result = mSubRegistryItem.emplace(std::make_pair(ItemName, Kratos::make_unique<TItemType>(ItemName, std::forward<TArgumentsList>(Arguments)...)));
-        KRATOS_ERROR_IF_NOT(insert_result.second) << "Error in inserting " << ItemName << " in registry item with name " << this->Name() << std::endl;
+        KRATOS_ERROR_IF_NOT(insert_result.second) << "Error in inserting '" << ItemName << "' in registry item with name '" << this->Name() << "'." << std::endl;
         return *insert_result.first->second;
     }
 
-
-
     ///@}
     ///@name Access
     ///@{
 
+    SubRegistryItemType::iterator begin();
+
+    SubRegistryItemType::const_iterator cbegin() const;
+
+    SubRegistryItemType::iterator end();
+
+    SubRegistryItemType::const_iterator cend() const;
+
+    key_return_iterator key_begin();
+
+    key_return_iterator key_end();
```

Comment:
```key_return_iterator```, ```key_begin```, and ```key_end``` are not standard C++ names, so they should follow the Kratos naming conventions. Other than that, ```key_return_iterator``` never mutates the ```map``` it iterates on so these functions can be ```const```.
```suggestion
    KeyReturnIterator KeyBegin() const;

    KeyReturnIterator KeyEnd() const;
```

Reply:
Fair enough. About the class, I'd call it `KeyReturnConstIterator`. Concerning the method names, I suggest calling them  `KeyConstBegin` and `KeyConstEnd`. I think this makes clear that we're working with a const iterator.